aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -840,6 +845,108 @@ Tensor index_select_cuda(const Tensor& self, int64_t dim, const Tensor& index) {
+  auto first = thrust::device_ptr<scalar_t>(self_.data_ptr<scalar_t>());
+  auto num_nonzeros_h = thrust::transform_reduce(
+                  thrust::cuda::par(thrust_allocator).on(stream),
+                  first, first + N, NonZeroOp<scalar_t>(), int(0), thrust::plus<int>());
+#else
+  size_t temp_storage_bytes=0;
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto num_nonzeros = allocator.allocate(sizeof(int));
+  cub::TransformInputIterator<bool, NonZeroOp<scalar_t>, scalar_t*> itr(self_.data_ptr<scalar_t>(), NonZeroOp<scalar_t>());
+  cub::DeviceReduce::Sum(nullptr, temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  auto temp_storage = allocator.allocate(temp_storage_bytes);
+  cub::DeviceReduce::Sum(temp_storage.get(), temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  int num_nonzeros_h;
+  C10_CUDA_CHECK(cudaMemcpyAsync(&num_nonzeros_h, num_nonzeros.get(), sizeof(int), cudaMemcpyDeviceToHost, stream));
+  //need to synchronize to make sure data is available on the host
+  C10_CUDA_CHECK(cudaStreamSynchronize(stream));
+#endif
+  //expected output size is num_nonzeros x ndim
+  //we are producing output with size {num_nonzeros, ndim} and strides {num_nonzeros, 1} (that is, transposed ndim x num_nonzeros output)
+  //we are able to directly use passed output with this size and strides, and we can also (per contract)
+  //resize passed output with incorrect sizes anyway we want.
+  //However, out with correct sizes and incorrect strides will have to be copied to from the intermediate we've produced.
+  bool need_to_copy = out.dim() == 2 && out.sizes()[0] == num_nonzeros_h && out.sizes()[1] == self.dim() && !out.t().is_contiguous();
+  at::Tensor out_temp = need_to_copy ?
+    at::native::empty_cuda({self.dim(), num_nonzeros_h}, out.options()) :
+    out.resize_({self.dim(), num_nonzeros_h});
+  //Scalars are expected to produce output of size (1,0), so we can't write to it
+  if (self.dim() > 0) {
+#ifdef __HIP_PLATFORM_HCC__
+    thrust::counting_iterator<int64_t> counting_itr(0);
+    thrust::copy_if(thrust::cuda::par(thrust_allocator).on(stream),
+    counting_itr, counting_itr+N, first, thrust::device_ptr<int64_t>(out_temp.data_ptr<int64_t>()),
+    NonZeroOp<scalar_t>());
+#else
+    cub::CountingInputIterator<int64_t> counting_itr(0);
+    temp_storage_bytes = 0;
+    cub::DeviceSelect::Flagged(nullptr, temp_storage_bytes, counting_itr, itr,
+      out_temp.data_ptr<int64_t>(), (int*)num_nonzeros.get(), N, stream);
+    temp_storage = allocator.allocate(temp_storage_bytes);
+    cub::DeviceSelect::Flagged(temp_storage.get(), temp_storage_bytes, counting_itr, itr,
+      out_temp.data_ptr<int64_t>(), (int*)num_nonzeros.get(), N, stream);
+#endif
+    if (num_nonzeros_h > 0 && self.dim() > 1){
+        int64_t div = 1;
+#ifndef __HIP_PLATFORM_HCC
+        auto thrust_allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
```

Comment:
Maybe unconditionally create the allocator at the beginning of the function to avoid this ifndef? Or is it important that it is only created as late as possible?

Reply:
It's not important that it's created late, but without HIP for 1d tensors it is not needed at all, so I avoid construction/destruction.